### PR TITLE
Feature/2374/token resource it

### PR DESCRIPTION
**Pull request by garyluu**
Two issues in one PR because they affected one another

For #2374 

Was failing because there are issues with Java 11, Powermock, and Jackson afterburner.
There was a quick fix for TokenResourceIT by completely avoiding the use of Jackson Afterburner, but decided not to because maintaining static powermock was a pain.

Using Hoverfly to intercept GitHub and Google calls instead which required changing the GoogleNetHttpTransport to NetHttpTransport because it needs to use the default SSLContext (not ideal but oh well).  Added the Hoverfly certificate to LocalTrustStore which is used in the non-confidential tests (likely to always be integration) to always trust Hoverfly.

Added a new profile that's non-confidential-tests which are integration tests that do not require confidential files (possibly due to mocking all external HTTP calls).  Because of this, It's intended to be run on any CI (namely CircleCI).

Also did some refactoring for the CommonTestUtilities so different Dropwizard configuration files can be used with non-confidential test data.

Avoided making too many TokenResourceIT test changes

For #2479 
Updating okhttp to 3.14.2 fixes it but OkUrlFactory was completely removed which was needed for github-api caching (tests fail otherwise).  https://github.com/kohsuke/github-api/issues/506 was made but doesn't look like it's resolved yet.  okhttp suggested using the obsolete factory https://github.com/square/okhttp/blob/master/CHANGELOG.md#version-3140 if absolutely necessary.  So using that until github-api issue is resolved.

### CODE REVIEW

**Comment by garyluu** [comment]
Didn't touch dockstore-jetty, don't know why snyk is complaining

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/ExtendedUserData.java [hunk 1]
```diff
@@ -20,25 +20,35 @@
...
     private boolean canChangeUsername;
 
     public ExtendedUserData(User user, PermissionsInterface authorizer) {
         Hibernate.initialize(user.getEntries());
-        this.canChangeUsername = user.getEntries().stream().noneMatch(Entry::getIsPublished) && !authorizer.isSharing(user) && user.getOrganizations().size() == 0;
+        try {
+            this.canChangeUsername = user.getEntries().stream().noneMatch(Entry::getIsPublished) && !authorizer.isSharing(user)
+                    && user.getOrganizations().size() == 0;
+        } catch (Exception e) {
+            LOG.error("SAM is improperly configured.");
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/ExtendedUserData.java]: Log the exception as a second param.

File: .circleci/config.yml [hunk 2]
```diff
@@ -38,6 +38,15 @@ jobs:
           name: run the actual tests
           command: |
             mvn --batch-mode clean install -Punit-tests
+      - run:
+          name: run the non-confidential integration tests
+          command: |
```
Comment (denis-yuen) [inline .circleci/config.yml]: Can avoid `cd`ing around with -pl
https://stackoverflow.com/questions/1114026/maven-modules-building-a-single-specific-module

File: .circleci/config.yml [hunk 3]
```diff
@@ -38,6 +38,15 @@ jobs:
...
+            # Adding all "normal" certs into this local one that has the Hoverfly cert (instead of adding Hoverfly cert to the global one)
```
Comment (denis-yuen) [inline .circleci/config.yml]: I think you mentioned this verbally in the demo, but for posterity, code comment on the downside of adding the hoverfly cert globally

File: dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java [hunk 4]
```diff
@@ -53,23 +55,36 @@
...
      */
-    public static final String CONFIDENTIAL_CONFIG_PATH = ResourceHelpers.resourceFilePath("dockstoreTest.yml");
+    public static final String CONFIDENTIAL_CONFIG_PATH = getConfidentialConfigPath();
     static final String DUMMY_TOKEN_1 = "08932ab0c9ae39a880905666902f8659633ae0232e94ba9f3d2094cb928397e7";
 
     private CommonTestUtilities() {
 
     }
 
+    private static String getConfidentialConfigPath() {
```
Comment (denis-yuen) [inline dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java]: In this case, I'd prefer the static initialization block right after CONFIDENTIAL_CONFIG_PATH to emphasize both the `final` and that this shouldn't be called again
https://stackoverflow.com/questions/9379426/java-when-is-a-static-initialization-block-useful

File: dockstore-integration-testing/src/test/java/io/dockstore/common/Hoverfly.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    private static String getFakeAccessToken(String suffix) {
+        return "fakeAccessToken" + suffix;
+    }
+
+    private static Tokeninfo getFakeTokeninfo(String email) {
+        Tokeninfo tokeninfo = new Tokeninfo();
+        tokeninfo.setAccessType("offline");
+        tokeninfo.setAudience("<fill me in>");
+        tokeninfo.setEmail(email);
+        tokeninfo.setExpiresIn(9001);
```
Comment (denis-yuen) [inline dockstore-integration-testing/src/test/java/io/dockstore/common/Hoverfly.java]: What's the unit here and why 9001? (code comment)

File: dockstore-integration-testing/src/test/java/io/dockstore/models/Satellizer.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
+/*
+ *
+ *  *    Copyright 2019 OICR
+ *  *
+ *  *    Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (denis-yuen) [inline dockstore-integration-testing/src/test/java/io/dockstore/models/Satellizer.java]: Some odd double-commenting going on?

File: pom.xml [hunk 7]
```diff
@@ -1333,6 +1346,7 @@
                 <skipTests>true</skipTests>
                 <skipITs>false</skipITs>
                 <skipClientITs>false</skipClientITs>
+                <argLine>-Xmx2048m</argLine>
```
Comment (denis-yuen) [inline pom.xml]: Is this necessary if the new test is being run in non-confidential-tests instead?

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/ExtendedUserData.java [hunk 8]
```diff
@@ -20,25 +20,35 @@
...
     private boolean canChangeUsername;
 
     public ExtendedUserData(User user, PermissionsInterface authorizer) {
         Hibernate.initialize(user.getEntries());
-        this.canChangeUsername = user.getEntries().stream().noneMatch(Entry::getIsPublished) && !authorizer.isSharing(user) && user.getOrganizations().size() == 0;
+        try {
+            this.canChangeUsername = user.getEntries().stream().noneMatch(Entry::getIsPublished) && !authorizer.isSharing(user)
+                    && user.getOrganizations().size() == 0;
+        } catch (Exception e) {
+            LOG.error("SAM is improperly configured.", e);
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/ExtendedUserData.java]: Missed this, why is SAM changing and throwing exceptions?

File: pom.xml [hunk 9]
```diff
@@ -44,7 +44,7 @@
         <postgresql.version>42.2.5</postgresql.version>
         <mockito.version>2.26.0</mockito.version>
         <cwlavro.version>2.0.4.2</cwlavro.version>
-        <okhttp.version>3.9.0</okhttp.version>
```
Comment (denis-yuen) [inline pom.xml]: This is more of a general comment, but noticed here that the powermock dependency isn't being deleted. 
Is it being used anywhere else?

File: dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java [hunk 10]
```diff
@@ -78,17 +93,39 @@ public static void dropAndRecreateNoTestData(DropwizardTestSupport<DockstoreWebs
...
         List<String> migrationList = Arrays.asList("1.3.0.generated", "1.3.1.consistency", "test", "1.4.0", "1.5.0", "test_1.5.0", "1.6.0", "1.7.0");
-        runMigration(migrationList, application, CONFIDENTIAL_CONFIG_PATH);
+        runMigration(migrationList, application, dropwizardConfigurationFile);
+    }
+
+    /**
+     * Shared convenience method
+     * @return
+     */
+    public static ApiClient getWebClient(boolean authenticated, String username) {
```
Comment (denis-yuen) [inline dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java]: Is this being moved from somewhere?
I don't see a corresponding delete anywhere

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/TokenResourceIT.java [hunk 11]
```diff
@@ -405,17 +367,15 @@ private void addUnavailableGoogleTokenToGitHubUser() {
...
-        mockGitHub(CUSTOM_USERNAME1);
+    private void addUnavailableGitHubTokenToGoogleUser() {
         TokensApi otherUserTokensApi = new TokensApi(getWebClient(true, GOOGLE_ACCOUNT_USERNAME2));
         try {
-            otherUserTokensApi.addGithubToken("potato");
+            otherUserTokensApi.addGithubToken(getFakeCode(SUFFIX1));
             Assert.fail();
         } catch (ApiException e){
             Assert.assertEquals(HttpStatus.SC_INTERNAL_SERVER_ERROR, e.getCode());
-            Assert.assertTrue(e.getMessage().contains("already exists"));;
```
Comment (denis-yuen) [inline dockstore-integration-testing/src/test/java/io/dockstore/webservice/TokenResourceIT.java]: Amusing, how did you notice this?

File: pom.xml (hunk 7 above)
Comment (garyluu) [reply pom.xml]: Yes

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/ExtendedUserData.java (hunk 8 above)
Comment (garyluu) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/core/ExtendedUserData.java]: Was testing without SAM properly configured in my Dropwizard configuration file and it gave a confusing message.

File: pom.xml (hunk 9 above)
Comment (garyluu) [reply pom.xml]: MockedIT

Comment (denis-yuen) [reply pom.xml]: Sad.
Guess we'll need a follow-up ticket, but later.

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/TokenResourceIT.java (hunk 11 above)
Comment (garyluu) [reply dockstore-integration-testing/src/test/java/io/dockstore/webservice/TokenResourceIT.java]: FindBugs in the IDE.  There a green checkmark when everything is ok.

**Pull request merged**